### PR TITLE
Add channel handler for server side pipelining.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

HTTP pipelining can be tricky to handle properly on the server side.
In particular, it's very easy to write out of order or inconsistently
mutate state. Users often need help to handle this appropriately.

Modifications:

Added a HTTPServerPipelineHandler that only lets one request through
at a time.

Result:

Better servers that are more able to handle HTTP pipelining

### CODE REVIEW

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+//
+//  HTTPServerPipelineHandler.swift
+//  NIOHTTP1
+//
+//  Created by Cory Benfield on 01/03/2018.
+//
+
+import NIO
+
+/// A `ChannelHandler` that handles HTTP pipelining by buffering inbound data until a
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: @Lukasa maybe we want to add a link to: https://tools.ietf.org/html/rfc7230#section-6.3.2 just to add a bit more context ?

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+/// response has been sent.
+///
+/// This handler ensures that HTTP server pipelines only process one request at a time.
+/// This is the safest way for pipelining-unaware code to operate, as it ensures that
+/// mutation of any shared server state is not parallelised, and that responses are always
+/// sent for each request in turn. In almost all cases this is the behaviour that a
+/// pipeline will want.
+public class HTTPServerPipelineHandler: ChannelInboundHandler, ChannelOutboundHandler {
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: nit: final ?

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+        if case .end = self.unwrapInboundIn(data) {
+            // New request is complete. We don't want any more data from now on.
+            self.responsePending = true
+        }
+        ctx.fireChannelRead(data)
+    }
+
+    public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        precondition(self.responsePending, "Unexpected data flow")
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: @Lukasa should this be a `assert` ?

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    /// Whether the channel pipeline is currently processing a response.
+    private var responsePending = false
+
+    /// While we're waiting to send the response we don't read from the socket.
+    /// This keeps track of whether we need to call read() when we've send our response.
+    private var readPending = false
+
+    /// The buffered HTTP requests that are not going to be addressed yet.
+    private var requestBuffer: [NIOAny] = []
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: I wonder if we want to use a `CircularBuffer` here ?. This sounds like a better fit in terms of overhead. WDYT ?

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    /// A response has been sent: deliver all pending requests and
+    /// mark the channel ready to handle more requests.
+    private func deliverPendingRequests(ctx: ChannelHandlerContext) {
+        self.responsePending = false
+        if self.requestBuffer.count > 0 {
+            let reads = self.requestBuffer
+            self.requestBuffer = []
+
+            reads.forEach { self.channelRead(ctx: ctx, data: $0) }
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: See above ... If we switch to `CircularBuffer` we can just empty it and there is nothing left.

**Comment by normanmaurer** [comment]
@Lukasa also I think we want to use this handler in our `HTTP1` example

**Comment by Lukasa** [comment]
Cool, updated. Time to rereview!

File: Sources/NIOHTTP1Server/main.swift [hunk 6]
```diff
@@ -399,6 +399,8 @@ let bootstrap = ServerBootstrap(group: group)
     // Set the handlers that are applied to the accepted Channels
     .childChannelInitializer { channel in
         channel.pipeline.addHTTPServerHandlers().then {
+            channel.pipeline.add(handler: HTTPServerPipelineHandler())
```
Comment (weissi) [inline Sources/NIOHTTP1Server/main.swift]: I think `addHTTPServerHandlers` should have some parameter of whether to insert `HTTPServerPipelineHandler` automatically

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+        if case .end = self.unwrapInboundIn(data) {
+            // New request is complete. We don't want any more data from now on.
+            self.state.requestEndReceived()
+        }
+        ctx.fireChannelRead(data)
+    }
+
+    public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        assert(self.state != .requestEndPending, "Unexpected data flow")
```
Comment (weissi) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: if I see this correctly, this happens when a second response is coming in end we're waiting for the previous request's `.end`. Could we have a better message here?

**Comment by normanmaurer** [comment]
This will fix https://github.com/apple/swift-nio/issues/43

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+///
+/// This handler ensures that HTTP server pipelines only process one request at a time.
+/// This is the safest way for pipelining-unaware code to operate, as it ensures that
+/// mutation of any shared server state is not parallelised, and that responses are always
+/// sent for each request in turn. In almost all cases this is the behaviour that a
+/// pipeline will want.
+///
+/// See https://tools.ietf.org/html/rfc7230#section-6.3.2 for more details on safely
+/// handling HTTP pipelining.
+public final class HTTPServerPipelineHandler: ChannelInboundHandler, ChannelOutboundHandler {
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: now that we have `ChannelDuplexHandler` please use it :)

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        while self.state != .responseEndPending, let read = self.requestBuffer.first {
+            _ = self.requestBuffer.removeFirst()
+            self.channelRead(ctx: ctx, data: read)
+            deliveredRead = true
+        }
+
+        if deliveredRead {
+            ctx.fireChannelReadComplete()
+        }
+    }
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: @Lukasa do we need to do something "smart" for the buffered writes if `close(ctx, mode, promise)` is called ? Maybe its ok to just drop these on the floor tho

Comment (Lukasa) [reply Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: I don't think so. If we got a close there's not really anything smart we can do with them anyway.

File: Sources/NIOHTTP1Server/main.swift (hunk 6 above)
Comment (Lukasa) [reply Sources/NIOHTTP1Server/main.swift]: Do you mind if we open an issue to track this work and I deal with it in another PR? Doing it well requires a fairly substantial rewrite of that code, and I don't want to make this PR handle two jobs.

Comment (normanmaurer) [reply Sources/NIOHTTP1Server/main.swift]: @Lukasa works for me

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift (hunk 9 above)
Comment (normanmaurer) [reply Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: sounds good...

**Comment by normanmaurer** [comment]
@Lukasa can you rebase ?

File: Sources/NIOHTTP1Server/main.swift (hunk 6 above)
Comment (weissi) [reply Sources/NIOHTTP1Server/main.swift]: Works for me too but maybe we should tag 1.1.0 before this PR so we can sort out possible APIs before we then tag 1.2.0 which will contain this PR? But you two decide, I’ll be offline for most of the day and off next week

**Comment by weissi** [comment]
@Lukasa please run

    git mv IntegrationTests/tests_01_http/SKIP_test_13_http_pipelining.sh  IntegrationTests/tests_01_http/test_13_http_pipelining.sh

to enable the pipelining tes

**Comment by jordan2936** [comment]
Man I don’t know what that is 


Sent from Yahoo Mail for iPhone


On Friday, March 2, 2018, 6:56 PM, Johannes Weiß <notifications@github.com> wrote:


@Lukasa please run
git mv IntegrationTests/tests_01_http/SKIP_test_13_http_pipelining.sh  IntegrationTests/tests_01_http/test_13_http_pipelining.sh

to enable the pipelining tes

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
@tomerd Seems like the builder is keeping hold of state from #109 which is causing this build to fail. Want to dive into that?

**Comment by tomerd** [comment]
> @tomerd Seems like the builder is keeping hold of state from #109 which is causing this build to fail. Want to dive into that?

@Lukasa this is fixed now, but this is filling an integration test

```
00:03:49.294 Running test 'test_13_http_pipelining.sh'... FAILURE (1)
```

**Comment by Lukasa** [comment]
@tomerd Yeah, I suspected it would, thanks. :smile:

**Comment by Lukasa** [comment]
So the failure here is because netcat half-closes the TCP connection when it gets EOF, which confuses the crap out of our test server. I'm going to work on that, but it'll take a few patches because the test server doesn't currently handle keep-alive properly either.

**Comment by Lukasa** [comment]
Thank goodness, this now works. Please re-review.

File: Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+                return b1 == b2
+            case (.halfClose, .halfClose):
+                return true
+            default:
+                return false
+            }
+        }
+    }
+
+    public var reads: [Event] = []
```
Comment (weissi) [inline Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift]: should this really be publicly mutable?

Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift]: In the tests I don't think it matters.

Nope, didn't run perf tests, will do shortly.

**Comment by Lukasa** [comment]
Perf results:

#### Without pipeline handler

```
wrk -c 1k -d 1m -t 4 --latency http://localhost:8888
Running 1m test @ http://localhost:8888
  4 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    17.19ms    3.15ms  56.53ms   84.12%
    Req/Sec    13.83k   842.71    16.24k    80.12%
  Latency Distribution
     50%   17.69ms
     75%   18.28ms
     90%   19.58ms
     99%   22.45ms
  3302121 requests in 1.00m, 160.61MB read
  Socket errors: connect 0, read 918, write 0, timeout 0
Requests/sec:  55004.64
Transfer/sec:      2.68MB
```

With the pipelining handler

```
wrk -c 1k -d 1m -t 4 --latency http://localhost:8888
Running 1m test @ http://localhost:8888
  4 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    16.84ms    4.89ms 168.78ms   86.04%
    Req/Sec    13.57k     1.02k   17.11k    86.04%
  Latency Distribution
     50%   17.90ms
     75%   18.45ms
     90%   19.69ms
     99%   23.24ms
  3240724 requests in 1.00m, 157.62MB read
  Socket errors: connect 0, read 1131, write 0, timeout 0
Requests/sec:  53972.47
Transfer/sec:      2.63MB
```

With the pipelining handler, and actually pipelining:

```
wrk -c 1k -d 1m -t 4 --latency -s pipeline.lua http://localhost:8888
Running 1m test @ http://localhost:8888
  4 threads and 1000 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   317.96ms  210.25ms   1.99s    62.87%
    Req/Sec    47.34k    14.35k   96.72k    66.08%
  Latency Distribution
     50%  291.35ms
     75%  471.95ms
     90%  626.93ms
     99%  824.32ms
  11300061 requests in 1.00m, 549.61MB read
  Socket errors: connect 0, read 2381, write 70, timeout 467
Requests/sec: 188102.51
Transfer/sec:      9.15MB
```

The TL;DR here is that the pipelining handler slows the server down in the non-pipelined mode by a worst-case of about 2% (these tests do no meaningful CPU work in the HTTP server itself, so the performance delta here basically is the worst case).

I'd say that's a small price to pay for correctness. ;)

(BTW, the read errors here are probably because this server closes non-keepalive connections, and that the way it does so could easily lead to TCP RST rather than EOF. We should address that in a future patch.)

**Comment by weissi** [comment]
@Lukasa the perf results look awesome 👌

File: Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift (hunk 10 above)
Comment (weissi) [reply Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift]: sorry, missed that it was in the tests, please ignore

**Comment by normanmaurer** [comment]
@Lukasa awesome! I will do a last review pass later today

**Comment by normanmaurer** [comment]
Also the 2% may just be the effect of having another handler in the pipeline in general

File: Sources/NIOHTTP1/HTTPDecoder.swift [hunk 11]
```diff
@@ -115,9 +115,13 @@ public extension ChannelPipeline {
...
     ///     - first: Whether to add the HTTP server at the head of the channel pipeline,
-    ///              or at the tail.
+    ///         or at the tail.
+    ///     - supportPipelining: Whether your code can manually handle pipelined requests,
+    ///         or whether it needs requests serialized. If `false` (the default), this
+    ///         call will also insert a `HTTPServerPipelineHandler` to serialize HTTP
+    ///         requests.
     /// - returns: An `EventLoopFuture` that will fire when the pipeline is configured.
-    public func addHTTPServerHandlers(first: Bool = false) -> EventLoopFuture<Void> {
+    public func addHTTPServerHandlers(first: Bool = false, supportPipelining: Bool = false) -> EventLoopFuture<Void> {
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPDecoder.swift]: @Lukasa @weissi I guess this change is ok as we have a default param, correct ?

Comment (weissi) [reply Sources/NIOHTTP1/HTTPDecoder.swift]: @normanmaurer yes 👍

File: Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+        switch event {
+        case let evt as ChannelEvent where evt == ChannelEvent.inputClosed:
+            self.reads.append(.halfClose)
+        default:
+            ctx.fireUserInboundEventTriggered(event)
+        }
+    }
+}
+
+private final class SuppressingReadsHandler: ChannelOutboundHandler {
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift]: IMHO the name is confusing as it really does not suppress reads but just count these and forward the reads to the next handler. Maybe call it `ReadCountHandler` ?

File: Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+
+        // Now send the half-closure.
+        self.channel.pipeline.fireUserInboundEventTriggered(ChannelEvent.inputClosed)
+
+        // The half-closure should be delivered immediately.
+        XCTAssertEqual(self.readRecorder.reads,
+                       [.read(HTTPServerRequestPart.head(self.requestHead)),
+                        .read(HTTPServerRequestPart.end(nil)),
+                        .read(HTTPServerRequestPart.head(self.requestHead)),
+                        .read(HTTPServerRequestPart.end(nil)),                        .halfClose])
```
Comment (normanmaurer) [inline Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift]: remove `whitespaces` in front of `.halfClose`

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+                preconditionFailure("Received second request")
+            }
+        }
+    }
+
+    /// The events that this handler buffers while waiting for the server to
+    /// generate a response.
+    private enum BufferedEvent {
+        /// A channelRead event.
+        case read(NIOAny)
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: maybe call it `channelRead` to ensure this is not confused with `read` method in `ChannelOutboundHandler`. Just a suggestion tho

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+/// This is the safest way for pipelining-unaware code to operate, as it ensures that
+/// mutation of any shared server state is not parallelised, and that responses are always
+/// sent for each request in turn. In almost all cases this is the behaviour that a
+/// pipeline will want. This is achieved without doing too much buffering by preventing
+/// the `Channel` from reading from the socket until a complete response is processed,
+/// ensuring that a malicious client is not capable of overwhelming a server by shoving
+/// an enormous amount of data down the `Channel` while a server is processing a
+/// slow response.
+///
+/// See https://tools.ietf.org/html/rfc7230#section-6.3.2 for more details on safely
```
Comment (normanmaurer) [inline Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: Consider using markdown syntax for the link: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links

**Comment by Lukasa** [comment]
Ok, feedback applied.

File: Sources/NIOHTTP1/HTTPServerPipelineHandler.swift (hunk 14 above)
Comment (normanmaurer) [reply Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: @Lukasa did you see this and decide that its ok as it is ? Thats fine, just wanted to ensure you not missed it

Comment (Lukasa) [reply Sources/NIOHTTP1/HTTPServerPipelineHandler.swift]: Oh lol, I only did it for the *other* enum. Fixed.

**Pull request merged**